### DISCUSSION TITLE
One-To-Many relation giving unique constraint error

### DISCUSSION THREAD

**Discussion by buenofs** [prisma/prisma#13225] [Q&A]
Hey guys,

So I came across a really weird behaviour from my schema and I'm not quite sure if I'm doing this right. What happens is, I'm trying to create records on my 1:n table but everytime I try to add a new record with the same id`s, it returns a unique constraint error (but it doesn't have any unique constraint).

Here are my models:

```prisma
model Post {
  id        Int      @id @default(autoincrement())
  text      String   @db.VarChar(777)
  createdAt DateTime @default(now())
  user      User     @relation(fields: [userId], references: [id])
  userId    Int
  reposts   Repost[]
}

model Repost {
  text      String?  @db.VarChar(777)
  post      Post     @relation(fields: [postId], references: [id])
  postId    Int
  user      User     @relation(fields: [userId], references: [id])
  userId    Int
  createdAt DateTime @default(now())

  @@id([postId, userId])
}
```

And here is my update query:

```ts
await this.prisma.post.update({
        where: { id: Number(postId) },
        data: {
          reposts: {
            create: [{ text: text, user: { connect: { id: Number(userId) } } }],
          },
        },
      });
```

The error I'm getting: 

```
 Unique constraint failed on the fields: (`postId`,`userId`)
    at Object.request (x\node_modules\@prisma\client\runtime\index.js:45629:15)
    at PrismaService._request (x\node_modules\@prisma\client\runtime\index.js:46456:18)
    at PostService.repostOrQuotePost (x\src\post\post.service.ts:76:7)
    at PostController.repostOrQuotePost (x\src\post\post.controller.ts:40:12)
    at x\node_modules\@nestjs\core\router\router-execution-context.js:46:28
    at x\node_modules\@nestjs\core\router\router-proxy.js:9:17

```

Any particular reason behind this? Maybe I'm doing wrong? What I want is to be able to repost as many posts I need since this is a 1:n relation, it should'nt have a unique constraint since I never made any `@unique` definition at all.

**Comment by austincrim** [comment]
Hey @buenofs 👋,

Do you already have a `Repost` in your database with the same combination of `userId` and `postId`? You might be trying to create a duplicate record. All `@id` or `@@id` attributes create a `unique`-type constraint under the hood.

You could try using [`connectOrCreate`](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#connectorcreate) instead of `create` to protect against this error.

**Reply by richardwardza** [reply]
Hi @buenofs 

what I think may also work is changing `  @@id([postId, userId])` in your `Repost` model to `@@index([postId, userId])` then the database should still index the columns but not make then unique.

**Reply by buenofs** [reply]
> Hi @buenofs 
> 
> 
> 
> what I think may also work is changing `  @@id([postId, userId])` in your `Repost` model to `@@index([postId, userId])` then the database should still index the columns but not make then unique.
> 
> 

Yeah that was exactly what I ended up doing, I also added an id as primary key. 

Mistake on my part, thanks guys!

**Comment by lecheynata** [comment]
hi @richardwardza  I had a similar problem and the way to solve it was to add the `id Int @id @default(autoincrement)`and the multifield `@@id([id, postId, userId])` to the model, to eliminate the possibility of duplicity in the records since that is the reason for the constraint error.

**Reply by buenofs** [reply]
Didn't knew it could be done like this too, could you share the difference between your solution and using @index ?

**Comment by richardwardza** [comment]
Using `@@id` should set a multi-field primary key on your table. 
It looks to be that [the consequences of this are ](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#id-1)that when you're trying to find a record you need to provide all the records to search on:
```
model User {
  firstName String
  lastName  String
  email     String  @unique
  isAdmin   Boolean @default(false)

  @@id([firstName, lastName])
}

const user = await prisma.user.findUnique({
  where: {
    firstName_lastName: {
      firstName: 'Alice',
      lastName: 'Smith',
    },
  },
})
```

**Comment by sharukhkhanajm** [comment]
This works fine.

`@@unique([id, userId])`

If anyone wondering why that's not working then you need to run this command `npx prisma db push` and this should fix your issue :)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
